File: 프로그래머스/lv3/김미래/기지국 설치.cpp
#include <iostream>
#include <vector>
using namespace std;
int solution(int n, vector<int> stations, int w)
{
    int answer = 0;
    int range = 2*w + 1;
    int start = 1;
    
    for(int i=0; i<stations.size(); i++){
       int boundary = stations[i] - w;
        if(start < boundary){
            int leng = boundary - start;
            if(leng <= range){
                answer +=1;
            }else{
                answer += leng/range;
                if(leng % range != 0){
                    answer += 1;
                }
            }
        }
        
        start = stations[i] + w + 1;
    }
    
    if(start<=n){
        answer += (n - start + 1) / range;
        if((n-start+1) % range != 0){
            answer +=1;
        }
    }

    return answer;
}

File: 프로그래머스/lv3/이재표/기지국 설치.java
import java.util.*;
class Solution {
    public int solution(int n, int[] stations, int w) {
        int answer = 0;
        int end = 1; 
        int window = w * 2 + 1; 
        for (int i = 0; i < stations.length; i++) {
            if (stations[i] - w > end) {
                int size = stations[i] - w - end; 
                int a = size / window;
                int b = size % window;
                if (b > 0) {
                    answer += a + 1;
                } else {
                    answer += a;
                }
            }
            end = stations[i] + w + 1;
        }
        if (end <= n) {
            int size = n - end + 1;
            int a = size / window;
            int b = size % window;
            if (b > 0) {
                answer += a + 1;
            } else {
                answer += a;
            }
        }
        
        return answer;
    }
}
